File: develop/lab-7-end-to-end-app/lab7JavaLab/src/main/java/com/amazonaws/pollynotes/solution/SearchFunctionSolution.java
package com.amazonaws.pollynotes.solution;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.amazonaws.pollynotes.pojo.Note;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBQueryExpression;
import com.amazonaws.services.dynamodbv2.model.AttributeValue;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;

public class SearchFunctionSolution implements RequestHandler<Note, List<Note>> {

	/**
	 * PollyNotes-SearchFunction
	 * 
     * This lambda function is integrated with the following API methods:
     * /notes/search/GET
     * 
     * This function does the following:
     * 1. Takes a JSON payload from API gateway and converts it into a Note POJO {@link com.amazonaws.pollynotes.pojo.Note}
     * 2. Queries DynamoDB using the QueryAsync feature. It will use the userId for the Partition Key, but a Filter will need
     *    to be applied to do the search based on the note text.
     * 3. Returns the List of Note found
     *  
     * @param	note 	POJO of Note created from the JSON sent from API GW into Lambda
     * @param	context	Lambda context which can be used for logging
     * @return			A List of Note containing all of the items for the userId containing the note  text in DynamoDB
	 */
	@Override
	public List<Note> handleRequest(Note note, Context context) {
		// The note object contains the userId and a partial note sent from API GW in the /notes/search/GET
        // {
        //   "userId": "...",
        //   "note": "..."
        // }

		context.getLogger().log("Initiating PollyNotes-SearchFunction...");
		context.getLogger().log("Note received: {" + note.toString() + "}");
		
		// Create the DynamoDB client and mapper
		AmazonDynamoDB client = AmazonDynamoDBClientBuilder.standard().build();
		DynamoDBMapper mapper = new DynamoDBMapper(client);
		
		// Create the Expression Attribute Value HashMap for the value of the note text to search
		//  sent in the note in the request
		Map<String, AttributeValue> expressionAttributeValues = new HashMap<String, AttributeValue>();
		expressionAttributeValues.put(":v_note", new AttributeValue().withS(note.getNote()));
		
		// Create the Query Expression with the Hash/Partition Key set to the one set in the note in the request
		//  and with the filter expression set to the value of the note sent in the request
		DynamoDBQueryExpression<Note> queryExpression = new DynamoDBQueryExpression<Note>()
				.withHashKeyValues(note)
				.withFilterExpression("contains (note, :v_note)")
					.withExpressionAttributeValues(expressionAttributeValues);
		
		try {
			// Query DynamoDB based on the Query Expression and return the results
			List<Note> noteList = mapper.query(Note.class, queryExpression);
			
			// Log the Note(s) found
			for (Note n : noteList) {
				context.getLogger().log("Note found in DynamoDB: {" + n.toString() + "}");
			}
			
			// Return the results
			return noteList;
		} catch (Exception e) {
			// Error found, log the message
			context.getLogger().log("Exception found, returning empty list. Error: " + e.getMessage());
			
			// Return an empty list
			return new ArrayList<Note>();
		}
	}
}

File: develop/lab-7-end-to-end-app/lab7CSharpLab/SearchFunctionSolution/SearchFunctionSolution.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DataModel;
using Amazon.DynamoDBv2.DocumentModel;
using Amazon.Lambda.Core;
using System.Threading.Tasks;


// Assembly attribute to enable the Lambda function's JSON input to be converted into a .NET class.
[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.Json.JsonSerializer))]

namespace SearchFunctionSolution
{
    public class SearchFunctionSolution
    {
        /// <summary>
        /// PollyNotes-SearchFunction
        /// 
        /// This lambda function is integrated with the following API methods:
        /// /notes/search/GET
        /// 
        /// This function does the following:
        /// 
        /// 1. Takes a JSON payload from API gateway and converts it into a Note POCO <see cref="Note"/>
        /// 2. Queries DynamoDB using the QueryAsync feature. It will use the userId for the Partition Key, but a Filter will need
        ///    to be applied to do the search based on the note text.
        /// 3. Returns the List of Note found
        /// </summary>
        /// <param name="note">POCO of Note from the JSON payload from API GW to Lambda. Only userId and note will be set.</param>
        /// <param name="context">Lambda context</param>
        /// <returns>List of Note from the Query to DynamoDB based on the userId and the note</returns>
        public List<Note> FunctionHandler(Note note, ILambdaContext context)
        {
            // The note object contains the userId and a partial note sent from API GW in the /notes/search/GET
            // {
            //   "userId": "...",
            //   "note": "..."
            // }

            Console.WriteLine("Initiating PollyNotes-SearchFunction...");
            Console.WriteLine("Note received: " + JsonConvert.SerializeObject(note));

            // Create the DynamoDB client and the Context for Object Persistence Model
            AmazonDynamoDBClient client = new AmazonDynamoDBClient();
            DynamoDBContext ddbcontext = new DynamoDBContext(client);

            // Create a DynamoDB Operation Config to pass in a QueryFilter on the note
            //  Since the note attribute isn't a key, we need to use a QueryFilter
            //  The QueryFilter takes a list of ScanCondition
            DynamoDBOperationConfig opConfig = new DynamoDBOperationConfig()
            {
                QueryFilter =
                        new List<ScanCondition>()
                        {
                            new ScanCondition("note", ScanOperator.Contains, note.note)
                        }
            };

            // Use the QueryAsync method to issue a query to DynamoDB based on the userId and QueryFilter
            //  We are using an Async command due to the .netcore SDK which doesn't support sync invokes
            //  We are using the Note POCO defined with the Object Persistence Model so all the data
            //   will be mapped to the right entries. The Table name is also defined in the POCO
            AsyncSearch<Note> noteQuery = ddbcontext.QueryAsync<Note>(note.userId, opConfig);

            // Retrieve all of the notes based on the Query from DynamoDB and wait
            Task<List<Note>> searchTask = noteQuery.GetRemainingAsync();
            searchTask.Wait();

            // Once the list of Note has been fetched entirely, check for Exceptions.
            // If there are no Exceptions
            if (searchTask.Exception == null)
            {
                Console.WriteLine("Successfully executed QueryAsync with userId: " + note.userId + " and text containing: " + note.note);

                // Log the Notes found
                Console.WriteLine("Notes found in DynamoDB: " + JsonConvert.SerializeObject(searchTask.Result));

                // Return the list
                return searchTask.Result;
            }
            else
            {
                // There was an exception, log the entry data and the exception
                Console.WriteLine("Unable to QueryAsync with userId: " + note.userId + " and text containing: " + note.note);
                Console.WriteLine(searchTask.Exception);

                // Return an empty list
                return new List<Note>();
            }
        }

        /// <summary>
        /// TESTING ONLY - Run the project as a console application to test the <see cref="FunctionHandler"/>
        /// You should see a list of Note
        /// </summary>
        public static void Main()
        {
            string searchTest = @"
            {  
                ""userId"":""testuser"",
                ""note"": ""test""
            }";
            List<Note> noteList = new SearchFunctionSolution().FunctionHandler(JsonConvert.DeserializeObject<Note>(searchTest), null);
            Console.WriteLine(JsonConvert.SerializeObject(noteList));
        }
    }
}
